File: geode-stats/src/main/java/org/apache/geode/internal/cache/tier/sockets/CCUStats.java
package org.apache.geode.internal.cache.tier.sockets;

import org.apache.geode.distributed.internal.DistributionStats;
import org.apache.geode.distributed.internal.ServerLocation;

/**
 * Stats for a CacheClientUpdater. Currently the only thing measured are incoming bytes on the
 * wire
 *
 * @since GemFire 5.7
 */
public class CCUStats implements MessageStats {

  private StatisticsType type;
  private int messagesBeingReceivedId;
  private int messageBytesBeingReceivedId;
  private int receivedBytesId;

  private void initializeStats(StatisticsFactory factory) {
    type = factory.createType("CacheClientUpdaterStats", "Statistics about incoming subscription data",
        new StatisticDescriptor[] {
            factory.createLongCounter("receivedBytes",
                "Total number of bytes received from the server.", "bytes"),
            factory.createIntGauge("messagesBeingReceived",
                "Current number of message being received off the network or being processed after reception.",
                "messages"),
            factory.createLongGauge("messageBytesBeingReceived",
                "Current number of bytes consumed by messages being received or processed.",
                "bytes"),});
    receivedBytesId = type.nameToId("receivedBytes");
    messagesBeingReceivedId = type.nameToId("messagesBeingReceived");
    messageBytesBeingReceivedId = type.nameToId("messageBytesBeingReceived");
  }

  // instance fields
  private final Statistics stats;

  CCUStats(StatisticsFactory factory, ServerLocation location) {
    initializeStats(factory);
    // no need for atomic since only a single thread will be writing these
    this.stats = factory.createStatistics(type, "CacheClientUpdater-" + location);
  }

  public void close() {
    this.stats.close();
  }

  @Override
  public void incReceivedBytes(long v) {
    this.stats.incLong(receivedBytesId, v);
  }

  @Override
  public void incSentBytes(long v) {
    // noop since we never send messages
  }

  @Override
  public void incMessagesBeingReceived(int bytes) {
    this.stats.incInt(messagesBeingReceivedId, 1);
    if (bytes > 0) {
      this.stats.incLong(messageBytesBeingReceivedId, bytes);
    }
  }

  @Override
  public void decMessagesBeingReceived(int bytes) {
    this.stats.incInt(messagesBeingReceivedId, -1);
    if (bytes > 0) {
      this.stats.incLong(messageBytesBeingReceivedId, -bytes);
    }
  }

  /**
   * Returns the current time (ns).
   *
   * @return the current time (ns)
   */
  public long startTime() {
    return DistributionStats.getStatTime();
  }
}

File: geode-micrometer-stats/src/main/kotlin/org/apache/geode/statistics/cache/CCUStats.kt
package org.apache.geode.statistics.cache

import org.apache.geode.internal.cache.tier.sockets.MessageStats
import org.apache.geode.statistics.internal.micrometer.CounterStatisticMeter
import org.apache.geode.statistics.internal.micrometer.GaugeStatisticMeter
import org.apache.geode.statistics.internal.micrometer.MicrometerMeterGroup

/**
 * Stats for a CacheClientUpdater. Currently the only thing measured are incoming bytes on the
 * wire
 *
 * @since GemFire 5.7
 */
class CCUStats internal constructor(private val serverLocation: String) : MicrometerMeterGroup("CacheClientUpdaterStats-$serverLocation"), MessageStats {

    override fun getCommonTags(): Array<String> = arrayOf("serverLocation", serverLocation)

    private val clientUpdaterReceivedBytesMeter = CounterStatisticMeter("client.updater.received.bytes.count", "Total number of bytes received from the server.", unit = "bytes")
    private val clientUpdateMessagesReceivedMeter = GaugeStatisticMeter("client.updater.received.messages.count", "Current number of message being received off the network or being processed after reception.")
    private val clientUpdateMessagesReceivedBytesMeter = GaugeStatisticMeter("client.updater.received.messages.bytes", "Current number of bytes consumed by messages being received or processed.", unit = "bytes")

    override fun initializeStaticMeters() {
        registerMeter(clientUpdaterReceivedBytesMeter)
        registerMeter(clientUpdateMessagesReceivedMeter)
        registerMeter(clientUpdateMessagesReceivedBytesMeter)
    }

    override fun incReceivedBytes(bytes: Long) {
        clientUpdaterReceivedBytesMeter.increment(bytes)
    }

    override fun incSentBytes(v: Long) {
        // noop since we never send messages
    }

    override fun incMessagesBytesBeingReceived(bytes: Int) {
        clientUpdateMessagesReceivedBytesMeter.increment(bytes)
    }

    override fun decMessagesBytesBeingReceived(bytes: Int) {
        clientUpdateMessagesReceivedBytesMeter.decrement(bytes)
    }

    fun incMessagesBeingReceived(bytes: Int) {
        clientUpdateMessagesReceivedMeter.increment()
        incMessagesBytesBeingReceived(bytes)
    }

    fun decMessagesBeingReceived(bytes: Int) {
        clientUpdateMessagesReceivedMeter.decrement()
        decMessagesBytesBeingReceived(bytes)
    }
}
